iceoryx_posh/CMakeLists.txt

```diff
@@ -383,6 +436,10 @@ if(TOML_CONFIG)
     set(ROUDI_EXPORT iceoryx_posh_config iox-roudi)
 endif()
 
+if (JSON_CONFIG)
+    set(ROUDI_EXPORT ${ROUDI_EXPORT} iceoryx_posh_config_json iox-roudi-json)
```

Comment:
@elfenpiff unfortunately nobody addressed my concerns from here https://github.com/eclipse-iceoryx/iceoryx/issues/835#issuecomment-855870578

I don't think it is a good idea to have both in parallel. Having multiple ways to do the same thing confuses the user. We should decide on one default config file type and advertise that. If someone wants to use a different one, that should be a cmake option. But I feel especially with the config file we don't do us a favor if we have multiple config file types. How many project you know do that and don't use a 3rd party abstraction, e.g. `QSettings`?

Instead of having a `JSON_CONFIG` and a `TOML_CONFIG` I would rather have a `CONFIG_FILE` option which could either be set to `TOML`, `JSON` or `NONE`. This makes it easier to extend in the future in case we decide to do so. With this, there would be only one `iox-roudi`, independent from what is used as config file.

We just have to decide what we want to have as default. I would vote for `TOML` (config file format) since it has some advantages compared to `JSON` (data exchange format), e.g. it supports comments. See also [first result for `config file formats` on google](https://jhall.io/posts/best-config-file-formats/). If there are issues with `cpptoml`, there are other options we can switch to e.g. `tomlc99` which is validated against two big test suites.

Additionally, we have to decide what to do with the non-default config file type if we extend the options in the default config file type. Do we have to also add all the options to the non-default one? Would this be a release blocker? If yes, how about when we get a third config file upstreamed, e.g. `lua` and a fourth one, e.g. `yaml` and a fifth one, e.g. `inl` and a sixth one, e.g. `XML`?

The same also applies to our gateways. At the moment there is only the Cyclone DDS gateway but in case the Fast DDS gateway gets merged one day we have to ask the same questions.

Reply:
1. I support the idea with the cmake option `CONFIG_FILE` with the 3 values `TOML`, `JSON` and `NONE`.
2. I also vote for `TOML` as default option.

How should we proceed here with the consistent config options? For this we definitely require some kind of abstraction where maybe a unified abstract path for each setting is described like `general.mempool.chunk.count : uint32_t` (maybe as vector) which is then translated into a TOML or JSON entry by the actual config parser.
In my opinion it is important that we describe all the config paths/entries only once with their type and translate them to toml/json otherwise we will always have consistency issues since some entries are slightly different named or some forget to add an entry or miss-spelled it.
@oschuene @elBoberido what do you think about this approach? @oschuene do you have time to add such an abstraction?

Reply:
@elfenpiff @oschuene on a second thought, maybe we still need `JSON_CONFIG` and `TOML_CONFIG` in order to build both at the same time in the CI and run the tests, else we would need to introduce another job just for this. But I would still make it a cmake option to decide which config file type to use for `iox-roudi`.

Regarding the abstraction. I'm not sure I got what you mean :smile: